kernel/objects/memory-amd64/objects/PageMapAmd64.cc

```diff
@@ -212,99 +212,85 @@ namespace mythos {
   optional<void> PageMap::MprotectOp::applyFrame(PageTableEntry* table, size_t index)
   {
     MLOG_INFO(mlog::cap, "mprotect", DVAR(index), DVARhex(flags.value));
-    PageTableEntry pme = table[index]; // copy the entry for later compare_exchange!
-    ASSERT(pme.page); // otherwise we should not have ended up here
+    PageTableEntry pme = table[index].load(); // atomic load of the entry for later compare_exchange!
+    // no assert here because the table could have been modified since the previous check! 
+    if (!pme.present || !pme.page) THROW(Error::LOST_RACE);
     auto entry = pme.writeable(flags.writable && pme.configurable)
-      //.executeDisabled(!req.executable) // not working on KNC
+      //.executeDisabled(!req.executable) // TODO not working on KNC
       .writeThrough(flags.write_through)
       .cacheDisabled(flags.cache_disabled);
-    table[index].replace(pme, entry); // simply ignore lost races
-    RETURN(Error::SUCCESS);
-  }
-
-  template<>
-  optional<void> PageMap::operateFrame<1>(PageTableEntry* table, FrameOp& op)
-  {
-    size_t startIndex = (op.vaddr()/pageSize(1)) % 512; // the part inside current table
-    size_t numEntries = (op.vaddr()+op.vsize()+pageSize(1)-1)/pageSize(1) - op.vaddr()/pageSize(1);
-    if (numEntries > 512-startIndex) numEntries = 512-startIndex;
-    for (size_t i=0; i < numEntries; i++) {
-      op.level = 1;
-      ASSERT(Align4k::is_aligned(op.vaddr())); // checked in invoke*()
-      ASSERT(op.vsize() >= Align4k::alignment()); // checked in invoke*()
-      auto res = op.applyFrame(table, startIndex+i);
-      if (!res) return res;
-      op.moveForward(pageSize(1));
-    }
+    if (!table[index].replace(pme, entry)) THROW(Error::LOST_RACE); // TODO or simply ignore the lost race?
     RETURN(Error::SUCCESS);
   }
 
   template<size_t LEVEL>
   optional<void> PageMap::operateFrame(PageTableEntry* table, FrameOp& op)
   {
-    size_t startIndex = (op.vaddr()/pageSize(LEVEL)) % 512; // the part inside current table
+    // calculate the start inside the current table. The %num_caps(LEVEL) is correct for the recursive table walk because 
+    // we advance page by page. Just the begin of the walk has to check that it actually lies within the root table.
+    size_t startIndex = (op.vaddr()/pageSize(LEVEL)) % num_caps(LEVEL);
```

Comment:
Maybe use (byte) offset into the table's vaddress range instead of vaddr. This gets rid of the modulo and potentially much clearer to the reader.